test/visual/mpl/circuit/test_circuit_matplotlib_drawer.py

```diff
@@ -64,19 +68,87 @@ def cwd(path):
         os.chdir(oldpwd)
 
 
-class TestMatplotlibDrawer(QiskitTestCase):
+class TestCircuitMatplotlibDrawer(QiskitTestCase):
     """Circuit MPL visualization"""
 
     def setUp(self):
         super().setUp()
-        self.circuit_drawer = TestMatplotlibDrawer.save_data_wrap(
+        self.circuit_drawer = TestCircuitMatplotlibDrawer.save_data_wrap(
             _matplotlib_circuit_drawer, str(self)
         )
 
+        if not os.path.exists(FAILURE_DIFF_DIR):
+            os.makedirs(FAILURE_DIFF_DIR)
+
+        if not os.path.exists(RESULTDIR):
+            os.makedirs(RESULTDIR)
+
     def tearDown(self):
         super().tearDown()
         mpl_close("all")
 
+    @staticmethod
+    def _image_path(image_name):
+        return os.path.join(RESULTDIR, image_name)
+
+    @staticmethod
+    def _reference_path(image_name):
+        return os.path.join(TEST_REFERENCE_PATH, image_name)
+
+    @staticmethod
+    def _new_gray(size, color):
+        img = Image.new("L", size)
+        drawing = ImageDraw.Draw(img)
+        drawing.rectangle((0, 0) + size, color)
+        return img
+
+    @staticmethod
+    def _black_or_b(diff_image, image, reference, opacity=0.85):
+        """Copied from https://stackoverflow.com/a/30307875"""
+        thresholded_diff = diff_image
+        for _ in range(3):
+            thresholded_diff = ImageChops.add(thresholded_diff, thresholded_diff)
+        size = diff_image.size
+        mask = TestCircuitMatplotlibDrawer._new_gray(size, int(255 * (opacity)))
+        shade = TestCircuitMatplotlibDrawer._new_gray(size, 0)
+        new = reference.copy()
+        new.paste(shade, mask=mask)
+        if image.size != new.size:
+            image = image.resize(new.size)
+        if image.size != thresholded_diff.size:
+            thresholded_diff = thresholded_diff.resize(image.size)
+        new.paste(image, mask=thresholded_diff)
+        return new
+
+    @staticmethod
+    def _get_black_pixels(image):
+        black_and_white_version = image.convert("1")
+        black_pixels = black_and_white_version.histogram()[0]
+        return black_pixels
+
+    @staticmethod
+    def _similarity_ratio(current, expected, image_name):
+        diff_name = current.split(".")
+        diff_name.insert(-1, "diff")
+        diff_name = ".".join(diff_name)
+
+        current = Image.open(current)
+        expected = Image.open(expected)
+
+        diff = ImageChops.difference(expected, current).convert("L")
+
+        black_pixels = TestCircuitMatplotlibDrawer._get_black_pixels(diff)
+        total_pixels = diff.size[0] * diff.size[1]
+        ratio = black_pixels / total_pixels
+
+        if ratio != 1:
+            TestCircuitMatplotlibDrawer._black_or_b(diff, current, expected).save(
+                FAILURE_DIFF_DIR + FAILURE_PREFIX + image_name, "PNG"
+            )
+        else:
+            TestCircuitMatplotlibDrawer._black_or_b(diff, current, expected).save(diff_name, "PNG")
+        return ratio
```

Comment:
It's surprising to me that a function called "similarity_ratio" would have the very major side effect of writing out images into hard-coded paths.  I think it would be neater to put that part of the logic into an `assertImagesEqual` method or the like - the magic-number decision to say that the cut-off for acceptable similarity is "precisely equal" is currently in two places: hidden in this function, and the `assert` that follows the call of this function.

Reply:
While there is a cut off here, this does not count as a test failure, but does aggregate all non 100% matches serpately. Perhaps `NOT_IDENTICAL` would be a better name.

Changed `_similary_ratio` to `_save_diff` to more accurately reflect what the function is doing.